float32 r
float32 theta
float32 phi
